piggypal/Models/paypal_api_stub.py

```diff
@@ -0,0 +1,103 @@
+
+        Returns values as strings in a list. Does not alter class or object variables.
+        """
+        # TODO: once the app goes live, this function should lead to the Paypal Auth screen
+        # ! Have Privacy in Mind
+        return sandbox_client_id, sandbox_client_secret
+
+    def _getOauthToken(self, client_id, client_secret) -> dict:
+        """
+        Requests an OAuth2 Token for provided client id and secret from the Paypal API
+        """
+
+        client = BackendApplicationClient(client_id=client_id)
+        oauth = OAuth2Session(client=client)
+        token = oauth.fetch_token(
+            token_url=self.auth_url,
+            client_id=client_id,
+            client_secret=client_secret
+        )
+        return token
+ 
+    @cherrypy.tools.json_in()
+    #@cherrypy.tools.json_out()
+    @cherrypy.tools.accept(media='text/plain')
+    def GET(self, start_date:'Internet Date/Time Format', end_date:'Internet Date/Time Format', **request_kwargs:dict) -> dict:
+        """
+        Fetches Transaction History from PayPal API via OAuth2 communication
+
+        - `start_date` and `end_date` should be provided in Internet Date/Time Format (https://tools.ietf.org/html/rfc3339#section-5.6)
+        - Fine-tuning in requests can be done via options specified in https://developer.paypal.com/docs/api/sync/v1/.
+        """
+        if not (start_date and end_date):
+            data = cherrypy.request.json
+            start_date = data["start-date"]
+            end_date = data["end_date"]
+
+        if not (dt_pattern.match(start_date) and dt_pattern.match(end_date)):
+            raise ValueError("start_date or end_date are not in the right format.")
+        try:
+            client = OAuth2Session(client_id=self._client_id, token=self._token)
+            response = client.get(self.transaction_url, params={"start_date":start_date, "end_date":end_date, **request_kwargs})
+        except TokenExpiredError as tee:
+            self._token = self._getOauthToken(self._client_id, self._client_secret)
+            client = OAuth2Session(client_id=self._client_id, token=self._token)
+            response = client.get(self.transaction_url, **request_kwargs)
+        #TODO format response as JSON
+        return response
+        
+
+    # TODO Define automatic Token refresh & update
+    # TODO Then update getTransactionHistory(...) and omit try...except...
+    # # (https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#third-recommended-define-automatic-token-refresh-and-update)
+    def _refreshToken(self):
+        pass
+
+if __name__ == '__main__': # pragma: no cover
+    conf = {
+        '/': {
+            'request.dispatch': cherrypy.dispatch.MethodDispatcher(),
+            #'tools.sessions.on': True,
+            'tools.response_headers.on': True,
+            'tools.response_headers.headers': [('Content-Type', 'text/plain')],
+        }
+    }
+    cherrypy.quickstart(Paypal_stub(), '/', conf)
```

Comment:
Is a unique base path identifier necessary for use in sessions?

Reply:
I am not sure what you mean... Please elaborate.

Reply:
Resolved via Discord.